### PR TITLE
[core] Introduce get taxonomy(ies) implementation in the content client

### PR DESCRIPTION
**Pull request by MenKNas**
- [ ] This PR follows the [Contribution Guide](https://github.com/Sitecore/content-sdk/blob/dev/CONTRIBUTING.md)
- [x] Changelog updated
- [ ] Upgrade guide entry added

## Description / Motivation
JSS-5343 -> This PR adds the `getTaxonomy` and `getTaxonomies` methods to the Content Client, providing the ability to retrieve taxonomy data from Content Services. It also introduces and exports the related queries and response types 

## Testing Details
- [x] Unit Test Added
- [x] Manual Test/Other

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -14,6 +14,10 @@ Our versioning strategy is as follows:
 
 ### 🎉 New Features & Improvements
 
+* `[core]` Introduced get taxonomy/taxonomies functionality on content client ([#99](https://github.com/Sitecore/content-sdk/pull/99)):
```
Comment (illiakovalenko) [inline CHANGELOG.md]: Can we group "content" related features under one entry? 
Let's leave exactly "[Name TBD]", naming is not yet decided
e.g.
* [Name TBD] support:
  * `[core]` Introduced get taxonomy/taxonomies functionality on content client
  * `[core]` Introduced get locales functionality on content client

File: packages/core/src/content/content-client.ts [hunk 2]
```diff
@@ -129,4 +137,183 @@ export class ContentClient {
...
   }
+
+  /**
+   * Retrieves all available taxonomies.
+   *
+   * This method supports optional pagination for both taxonomies and their terms.
+   * - If `taxonomyPageSize` is provided, it limits the number of taxonomies returned per request.
+   * - If `termsPageSize` is provided, it limits the number of terms returned per taxonomy per request.
+   * - If either is omitted, the API defaults are used (returns all results in a single request if no pagination is required).
+   *
```
Comment (illiakovalenko) [inline packages/core/src/content/content-client.ts]: Let's remove this line, it shows warning
```suggestion
```

File: packages/core/src/content/content-client.ts [hunk 3]
```diff
@@ -129,4 +137,183 @@ export class ContentClient {
...
     return response.manyLocale;
   }
+
+  /**
+   * Retrieves all available taxonomies.
+   *
+   * This method supports optional pagination for both taxonomies and their terms.
+   * - If `taxonomyPageSize` is provided, it limits the number of taxonomies returned per request.
+   * - If `termsPageSize` is provided, it limits the number of terms returned per taxonomy per request.
+   * - If either is omitted, the API defaults are used (returns all results in a single request if no pagination is required).
```
Comment (illiakovalenko) [inline packages/core/src/content/content-client.ts]: All these descriptions should be a part of each appropriate parameter below, it makes sense to group.

File: packages/core/src/content/content-client.ts [hunk 4]
```diff
@@ -129,4 +137,183 @@ export class ContentClient {
...
+
+    return fetchTaxonomiesPage();
+  }
+
+  /**
+   * Retrieves a single taxonomy by its ID.
+   *
+   * This method supports optional pagination for the terms within the taxonomy.
+   * - If `termsPageSize` is provided, it limits the number of terms returned per request.
+   * - If omitted, the API default is used (returns all terms in a single request).
```
Comment (illiakovalenko) [inline packages/core/src/content/content-client.ts]: Should be a part of parameter's description

File: packages/core/src/content/content-client.ts [hunk 5]
```diff
@@ -129,4 +137,183 @@ export class ContentClient {
...
+   *
+   * The returned taxonomy object includes the system metadata, the initial set of terms,
+   * and a `fetchNext` function for retrieving additional terms if more are available.
```
Comment (illiakovalenko) [inline packages/core/src/content/content-client.ts]: It can be a part of "returns" statement below

File: packages/core/src/content/content-client.ts [hunk 6]
```diff
@@ -129,4 +137,183 @@ export class ContentClient {
...
+  ): Promise<ContentItemList<Taxonomy>> {
+    const fetchTaxonomiesPage = async (after: string = ''): Promise<ContentItemList<Taxonomy>> => {
+      debug.content(
+        'Fetching taxonomies (pageSize: %s, after: %s)',
+        taxonomyPageSize ?? 'API Default',
+        after
+      );
+
+      const variables: Record<string, any> = {
+        after: after || '',
```
Comment (illiakovalenko) [inline packages/core/src/content/content-client.ts]: The fallback is defined on parameter level above
```suggestion
        after,
```

File: packages/core/src/content/content-client.ts [hunk 7]
```diff
@@ -129,4 +137,183 @@ export class ContentClient {
...
+            debug.content(
+              'Fetching terms for taxonomy %s (pageSize: %s, after: %s)',
+              taxonomy.system.id,
+              termsPageSize ?? 'API Default',
+              termsAfter
+            );
+
+            const termVariables: Record<string, any> = {
+              after: after || '',
+              termsAfter: termsAfter || '',
```
Comment (illiakovalenko) [inline packages/core/src/content/content-client.ts]: The fallback is defined on parameter level above
```suggestion
              after,
              termsAfter,
```

File: packages/core/src/content/taxonomies.ts [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+    results: {
+      /** The terms associated with the taxonomy, in a paginated format. */
+      terms: {
+        results: Term[];
+        cursor?: string | null;
+        hasMore: boolean;
+      };
+      /** The system metadata of the taxonomy. */
+      system: TaxonomySystem;
+    }[];
```
Comment (illiakovalenko) [inline packages/core/src/content/taxonomies.ts]: I suppose that "results" should represent _Taxonomy[]_ type, right?

File: packages/core/src/content/taxonomies.ts [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+ * Variables:
+ * - minimumPageSize: The number of taxonomies to retrieve per page.
+ * - termsPageSize: The number of terms to retrieve per page within each taxonomy.
+ * - after: The cursor for fetching the next page of taxonomies.
+ * - termsAfter: The cursor for fetching the next page of terms within a taxonomy.
+ */
+export const GET_TAXONOMIES_QUERY = `
+  query GetAllTaxonomies(
+    $minimumPageSize: Int
+    $termsPageSize: Int
```
Comment (illiakovalenko) [inline packages/core/src/content/taxonomies.ts]: Can we align naming of these variables, please? You will update it in methods as well
First one can be simplified and named as $pageSize so it's expected that it belongs to taxonomies, second can be left as it's $termsPageSize

File: packages/core/src/content/content-client.ts [hunk 10]
```diff
@@ -129,4 +137,183 @@ export class ContentClient {
...
+        termsAfter: '', // Always provide for API requirements
+      };
+
+      if (taxonomyPageSize !== undefined) {
+        variables.minimumPageSize = taxonomyPageSize;
+      }
+
+      if (termsPageSize !== undefined) {
+        variables.termsPageSize = termsPageSize;
+      }
```
Comment (illiakovalenko) [inline packages/core/src/content/content-client.ts]: Can we introduce a shared function to get variables? so it will reduce amount of code.
This function should be introduced at the start of `fetchTaxonomiesPage` function.
(pageSize, after - belongs to taxonomies variables)
You will be able to reuse it twice

```ts
const createVariables = (after = '', termsAfter = '') => ({
      pageSize,
      after,
      termsAfter,
      termsPageSize,
    });
```

File: packages/core/src/content/content-client.ts [hunk 11]
```diff
@@ -129,4 +137,183 @@ export class ContentClient {
...
+        id,
+        termsPageSize ?? 'API Default',
+        termsAfter
+      );
+
+      const variables: Record<string, any> = { id, termsAfter };
+
+      if (termsPageSize !== undefined) {
+        variables.termsPageSize = termsPageSize;
+      }
```
Comment (illiakovalenko) [inline packages/core/src/content/content-client.ts]: can it be set on object level? so we don't need to check additionally

**Review by illiakovalenko — CHANGES_REQUESTED** [review]
Looks really good!
Please, review my comments below and I will revisit again later

File: packages/core/src/content/taxonomies.ts (hunk 8 above)
Comment (MenKNas) [reply packages/core/src/content/taxonomies.ts]: This is response structure so it's ok to keep it as it is. Inside the client method it is converted into `Taxonomy[]`

File: CHANGELOG.md [hunk 12]
```diff
@@ -14,6 +14,12 @@ Our versioning strategy is as follows:
 
 ### 🎉 New Features & Improvements
 
+* `[core]` 
+  * `[Content - Name TBD] Support`:
```
Comment (illiakovalenko) [inline CHANGELOG.md]:
```suggestion
  * [Content - Name TBD] Support:
```

File: packages/core/src/content/content-client.ts [hunk 13]
```diff
@@ -129,4 +136,89 @@ export class ContentClient {
     const response = await this.get<LocalesQueryResponse>(GET_LOCALES_QUERY);
     return response.manyLocale;
   }
+
+  /**
+   * Retrieves all available taxonomies with optional pagination support.
+   * Note: Only taxonomies are paginated. Terms within each taxonomy are returned in full (no term pagination).
+   * @param {object} [options] - Optional pagination options.
+   * @param {number} [options.pageSize] - Optional. Limits the number of taxonomies returned per page. If not provided, defaults to Content Services API default (20).
```
Comment (illiakovalenko) [inline packages/core/src/content/content-client.ts]: For now let's leave it as TBD, we can't use this name
```suggestion
   * @param {number} [options.pageSize] - Optional. Limits the number of taxonomies returned per page. If not provided, defaults to [Name TBD] API default (20).
```

File: packages/core/src/content/content-client.ts [hunk 14]
```diff
@@ -129,4 +136,89 @@ export class ContentClient {
...
+   * @param {string} [options.after] - Optional. Cursor for pagination. Used to fetch the next page of taxonomies.
+   * @returns A promise that resolves to a paginated list of taxonomies, including pagination metadata (`hasMore`, `cursor`) for developer-controlled pagination.
+   */
+  async getTaxonomies(options?: {
+    pageSize?: number;
+    after?: string;
+  }): Promise<ContentItemList<Taxonomy>> {
+    debug.content(
+      'Fetching taxonomies (pageSize: %s, after: %s)',
```
Comment (illiakovalenko) [inline packages/core/src/content/content-client.ts]: Just to make wording consistent
```suggestion
      'Getting taxonomies (pageSize: %s, after: %s)',
```

File: packages/core/src/content/content-client.ts [hunk 15]
```diff
@@ -129,4 +136,89 @@ export class ContentClient {
...
+      options?.pageSize ?? 'API Default',
+      options?.after ?? ''
+    );
+
+    const variables: Record<string, any> = {};
+    if (options?.pageSize !== undefined) variables.pageSize = options.pageSize;
+
+    variables.after = options?.after ?? '';
```
Comment (illiakovalenko) [inline packages/core/src/content/content-client.ts]: Can we set props in variable, just for simplfication?
Can we set `pageSize` without checking for `undefined`?
```ts
    const variables: Record<string, any> = {
      pageSize: options?.pageSize,
      after: options?.after ?? '',
    };
```

File: packages/core/src/content/taxonomies.ts [hunk 16]
```diff
@@ -0,0 +1,196 @@
...
+ * In getTaxonomies, terms are always returned as a full list (no pagination).
+ * In getTaxonomy, terms may be paginated.
+ */
+export interface TermList {
+  /** The list of terms in the current (or full) page. */
+  results: Term[];
+  /** The cursor to fetch the next page of terms, if available. (Used only in getTaxonomy) */
+  cursor?: string | null;
+  /** Indicates whether more terms are available after the current page. (Used only in getTaxonomy) */
+  hasMore: boolean;
```
Comment (illiakovalenko) [inline packages/core/src/content/taxonomies.ts:24]: Do we suppose that `hasMore` is optional as well? I noticed you check whether it's present

File: packages/core/src/content/content-client.ts [hunk 17]
```diff
@@ -129,4 +136,89 @@ export class ContentClient {
...
+
+    const response = await this.get<TaxonomiesQueryResponse>(GET_TAXONOMIES_QUERY, variables);
+    const data = response?.manyTaxonomy;
+
+    return {
+      results: (data?.results ?? []).map((taxonomy) => ({
+        system: taxonomy.system,
+        terms: {
+          results: taxonomy.terms?.results ?? [],
+          cursor: undefined,
```
Comment (illiakovalenko) [inline packages/core/src/content/content-client.ts]: Can be removed
```suggestion
```

File: packages/core/src/content/taxonomies.ts [hunk 18]
```diff
@@ -0,0 +1,196 @@
...
+  /** The minimum page size used for fetching terms (if provided). (Used only in getTaxonomy) */
+  minimumPageSize?: number;
```
Comment (illiakovalenko) [inline packages/core/src/content/taxonomies.ts]: This one is not present, right?
```suggestion
```

File: packages/core/src/content/content-client.ts [hunk 19]
```diff
@@ -129,4 +136,89 @@ export class ContentClient {
...
+          hasMore: false,
+        },
+      })),
+      cursor: data?.cursor,
+      hasMore: data?.hasMore ?? false,
```
Comment (illiakovalenko) [inline packages/core/src/content/content-client.ts:167]: Does it mean that `hasMore` is optional? Should it be optional in our interfaces as well?

File: packages/core/src/content/content-client.ts [hunk 20]
```diff
@@ -129,4 +136,89 @@ export class ContentClient {
...
+    const response = await this.get<TaxonomiesQueryResponse>(GET_TAXONOMIES_QUERY, variables);
+    const data = response?.manyTaxonomy;
+
+    return {
+      results: (data?.results ?? []).map((taxonomy) => ({
+        system: taxonomy.system,
+        terms: {
+          results: taxonomy.terms?.results ?? [],
+          cursor: undefined,
+          hasMore: false,
```
Comment (illiakovalenko) [inline packages/core/src/content/content-client.ts]: Can be removed as well, if it's optional

File: packages/core/src/content/taxonomies.ts [hunk 21]
```diff
@@ -0,0 +1,196 @@
...
+};
+
+/**
+ * Represents a list of terms within a taxonomy.
+ * In getTaxonomies, terms are always returned as a full list (no pagination).
+ * In getTaxonomy, terms may be paginated.
+ */
+export interface TermList {
+  /** The list of terms in the current (or full) page. */
+  results: Term[];
```
Comment (illiakovalenko) [inline packages/core/src/content/taxonomies.ts:20]: Do we expect that `results` can be empty? In both Taxonomies and Terms

File: packages/core/src/content/content-client.ts [hunk 22]
```diff
@@ -129,4 +136,89 @@ export class ContentClient {
...
+    variables.after = options?.after ?? '';
+
+    const response = await this.get<TaxonomiesQueryResponse>(GET_TAXONOMIES_QUERY, variables);
+    const data = response?.manyTaxonomy;
+
+    return {
+      results: (data?.results ?? []).map((taxonomy) => ({
+        system: taxonomy.system,
+        terms: {
+          results: taxonomy.terms?.results ?? [],
```
Comment (illiakovalenko) [inline packages/core/src/content/content-client.ts]: since we don't support pagination for terms, probably it would be more user friendly if terms results can be set in `terms` property, right? I don't need to access `results` and especially `cursor` or `hasMore`

File: packages/core/src/content/content-client.ts [hunk 23]
```diff
@@ -129,4 +136,89 @@ export class ContentClient {
...
+    debug.content(
+      'Getting taxonomy for id: %s (termsPageSize: %s, termsAfter: %s)',
+      id,
+      terms?.pageSize ?? 'API Default',
+      terms?.after ?? ''
+    );
+
+    const variables: Record<string, any> = { id };
+    if (terms?.pageSize !== undefined) variables.termsPageSize = terms.pageSize;
+    if (terms?.after !== undefined) variables.termsAfter = terms.after;
```
Comment (illiakovalenko) [inline packages/core/src/content/content-client.ts]: Can it be simplified?
```suggestion
    const variables: Record<string, any> = {
      id,
      termsPageSize: terms?.pageSize,
      termsAfter: terms?.after,
    };
```

**Review by illiakovalenko — CHANGES_REQUESTED** [review]

File: packages/core/src/content/taxonomies.ts (hunk 16 above)
Comment (MenKNas) [reply packages/core/src/content/taxonomies.ts:24]: `hasMore` will always be present in the response so it should stay required in the type

File: packages/core/src/content/content-client.ts (hunk 19 above)
Comment (MenKNas) [reply packages/core/src/content/content-client.ts:167]: I have implemented this fallback option in the case `hasMore`  for some weird reason comes back as `undefined` or `null` from the API. In this case it would be better for anyone using our methods to have a default `false`  as the value of  `hasMore`

File: packages/core/src/content/taxonomies.ts (hunk 21 above)
Comment (MenKNas) [reply packages/core/src/content/taxonomies.ts:20]: It can be empty

File: packages/core/src/content/content-client.ts [hunk 24]
```diff
@@ -129,4 +135,83 @@ export class ContentClient {
...
+   * @returns A promise that resolves to an object containing taxonomies, their terms, and pagination info.
+   */
+  async getTaxonomies(options?: { pageSize?: number; after?: string }) {
+    debug.content(
+      'Getting taxonomies (pageSize: %s, after: %s)',
+      options?.pageSize ?? 'API Default',
+      options?.after ?? ''
+    );
+
+    const variables: Record<string, any> = {
```
Comment (illiakovalenko) [inline packages/core/src/content/content-client.ts]: `any` is not allowed to be used, you can just remove explicit type
```suggestion
    const variables = {
```

File: packages/core/src/content/content-client.ts [hunk 25]
```diff
@@ -129,4 +135,83 @@ export class ContentClient {
...
+    };
+  }): Promise<Taxonomy | null> {
+    debug.content(
+      'Getting taxonomy for id: %s (termsPageSize: %s, termsAfter: %s)',
+      id,
+      terms?.pageSize ?? 'API Default',
+      terms?.after ?? ''
+    );
+
+    const variables: Record<string, any> = {
```
Comment (illiakovalenko) [inline packages/core/src/content/content-client.ts]: any is not allowed to be used, you can just remove explicit type
```suggestion
    const variables = {
```

File: packages/core/src/content/content-client.test.ts [hunk 26]
```diff
@@ -226,4 +227,220 @@ describe('content-client', () => {
...
+      const callArgs = requestStub.getCall(0).args;
+      expect(callArgs[0]).to.equal(GET_TAXONOMIES_QUERY);
+      expect(callArgs[1]).to.deep.equal({ pageSize: 2, after: 'cursor-taxonomies' });
+      expect(result.results[0].system.name).to.equal('Taxonomy 2');
+      expect(result.results[0].terms[0].name).to.equal('Term 2');
+      expect(result.cursor).to.equal('another-cursor');
+      expect(result.hasMore).to.be.false;
+    });
+
+    it('should handle empty or null responses', async () => {
```
Comment (illiakovalenko) [inline packages/core/src/content/content-client.test.ts:329]: Let's put it under "describe" and add one extra test to cover `null | undefined` _terms.results_

File: packages/core/src/content/content-client.test.ts [hunk 27]
```diff
@@ -226,4 +227,220 @@ describe('content-client', () => {
...
+        termsPageSize: undefined,
+        termsAfter: undefined,
+      });
+      expect(result?.system.id).to.equal(taxonomyId);
+      expect(result?.terms.results[0].name).to.equal('Term 2');
+      expect(result?.terms.cursor).to.be.null;
+      expect(result?.terms.hasMore).to.be.false;
+    });
+
+    it('should handle null taxonomy response', async () => {
```
Comment (illiakovalenko) [inline packages/core/src/content/content-client.test.ts:508]: Let's put it under "describe" and add one extra test to cover `null | undefined` _terms.results_

**Review by illiakovalenko — APPROVED** [review]
Looks very good!
I left few minor comments, please, have a look
Feel free to merge then
